plugins/rts54hub/fu-rts54hub-rtd21xx-device.c

```diff
@@ -0,0 +1,618 @@
+
+	if (target_addr != self->target_addr) {
+		if (!fu_rts54hub_device_i2c_config (parent, target_addr, 1,
+						    FU_RTS54HUB_I2C_SPEED_200K,
+						    error))
+			return FALSE;
+		self->target_addr = target_addr;
+	}
+	if (!fu_rts54hub_device_i2c_write (parent, sub_addr, data, datasz, error)) {
+		g_prefix_error (error,
+				"failed to write I2C @0x%02x:%02x: ",
+				target_addr, sub_addr);
+		return FALSE;
+	}
+	g_usleep (I2C_DELAY_AFTER_SEND);
+	return TRUE;
+}
+
+static gboolean
+fu_rts54hub_rtd21xx_device_i2c_read (FuRts54hubRtd21xxDevice *self,
+				     guint8 target_addr, guint8 sub_addr,
+				     guint8 *data, gsize datasz,
+				     GError **error)
+{
+	FuRts54HubDevice *parent;
+
+	parent = fu_rts54hub_rtd21xx_device_get_parent (self, error);
+	if (parent == NULL)
+		return FALSE;
+	if (!fu_rts54hub_device_vendor_cmd (parent, 1, error))
+		return FALSE;
+	if (target_addr != self->target_addr) {
+		if (!fu_rts54hub_device_i2c_config (parent, target_addr, 1,
+						    FU_RTS54HUB_I2C_SPEED_200K,
+						    error))
+			return FALSE;
+		self->target_addr = target_addr;
+	}
+	if (!fu_rts54hub_device_i2c_read (parent, sub_addr, data, datasz, error)) {
+		g_prefix_error (error, "failed to read I2C: ");
+		return FALSE;
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_rts54hub_rtd21xx_device_read_status_raw (FuRts54hubRtd21xxDevice *self,
+					    guint8 *status,
+					    GError **error)
+{
+	guint8 buf[] = { 0x00 };
+	if (!fu_rts54hub_rtd21xx_device_i2c_read (self,
+						  UC_FOREGROUND_SLAVE_ADDR,
+						  UC_FOREGROUND_STATUS,
+						  buf, sizeof(buf),
+						  error))
+		return FALSE;
+	if (status != NULL)
+		*status = buf[0];
+	return TRUE;
+}
+
+static gboolean
+fu_rts54hub_rtd21xx_device_read_status_cb (FuDevice *device,
+					   gpointer user_data,
+					   GError **error)
+{
+	FuRts54hubRtd21xxDevice *self = FU_RTS54HUB_RTD21XX_DEVICE (device);
+	guint8 status = 0xfd;
+	if (!fu_rts54hub_rtd21xx_device_read_status_raw (self, &status, error))
+		return FALSE;
+	if (status == ISP_STATUS_BUSY) {
+		g_set_error (error,
+				FWUPD_ERROR,
+				FWUPD_ERROR_INTERNAL,
+				"status was 0x%02x", status);
+		return FALSE;
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_rts54hub_rtd21xx_device_read_status (FuRts54hubRtd21xxDevice *self,
+					guint8 *status,
+					GError **error)
+{
+	return fu_device_retry (FU_DEVICE (self),
+				fu_rts54hub_rtd21xx_device_read_status_cb,
+				4200, status, error);
+}
+
+static gboolean
+fu_rts54hub_rtd21xx_ensure_version_unlocked (FuRts54hubRtd21xxDevice *self,
+					     GError **error)
+{
+	guint8 buf_rep[7] = { 0x00 };
+	guint8 buf_req[] = { ISP_CMD_GET_FW_INFO };
+	g_autofree gchar *version = NULL;
+	if (!fu_rts54hub_rtd21xx_device_i2c_write (self,
+						   UC_FOREGROUND_SLAVE_ADDR,
+						   UC_FOREGROUND_OPCODE,
+						   buf_req, sizeof(buf_req),
+						   error)) {
+		g_prefix_error (error, "failed to get version number: ");
+		return FALSE;
+	}
+
+	/* wait for device ready */
+	g_usleep(300000);
```

Comment:
if this is waiting for a device to be ready, it should be a fu_retry with a timeout after a few tries instead of a magic one time value.

Reply:
IIRC from the VIA discussions, I think the i2c bus has to be quiet for this `ts`, and a retry wouldn't work.

Reply:
How do you make I2C bus quiet?  Anything else on the bus can be talking at this time?

Reply:
@AnyProblem got any more info?